### PR TITLE
Delay invoking clang version checks until a functional path is taken

### PR DESCRIPTION
**Pull request by kentfredric**
This allows avoiding large numbers of system calls to dynaload clang
to determine its version, when no action is performed, for example:

- When calling `--version` / `-V`
- When calling `--help`

This improves the raw responsiveness from:

  Before:
```
    time bindgen --help # 0.593s
    strace -cf bindgen --help # 83k syscalls, 64k to statx
```
  After:
```
    time bindgen --help # 0.004s
    strace -cf bindgen --help # 90 syscalls
```
However, it does mean that you can no longer obtain the discovered
clang version with:
```bash
RUST_LOG=info bindgen -V
```
But this may be remedied in a future commit.

Closes: https://github.com/rust-lang/rust-bindgen/issues/1736

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -46,9 +41,17 @@ pub fn main() {
     if expected_version.is_some() {
         assert_eq!(version.parsed, version.parsed);
```
Comment (emilio) [inline src/main.rs]: We should probably make his something like: `assert_eq!(version.parsed, expected_version)` or such, but that can be a separate PR.

**Comment by kentfredric** [comment]
> WIP?

Mostly because I wanted to work out if I needed to satisfy the "show clang version" in either of `-V` or `--version` as per my comments in https://github.com/rust-lang/rust-bindgen/issues/1736#issuecomment-590037571 

If these are not necessary or desired, then I won't push further on those aspects.

**Comment by emilio** [comment]
Yeah, I don't think that's needed. It could be done with a `log_enabled!` check or such, if needed, but it doesn't seem necessary IMHO.

**Comment by emilio** [comment]
Thank you!
